edgexfoundry/edgex-go
# feat(security): Implementation for generating and storing Consul tokens for EdgeX services

changes for `security-bootstrapper` on setupRegitrsyACL
 - add logic for deleting consul tokens for the 2nd time or later based on the consul-token files
 - add logic for generating consul tokens with go-mod-secret's API
 - add logic for re-set up the consul tokens for the 2nd time or later
 - add env. override `STAGEGATE_REGISTRY_ACL_TOKENBASEDIR` for snap on consul token base directory

Closes: #3306

Signed-off-by: Jim Wang <yutsung.jim.wang@intel.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x ] Tests for the changes have been added (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-go/blob/master/.github/Contributing.md.

## What is the current behavior?

N/A, this is new feature.

## Issue Number: #3306 


## What is the new behavior?
New feature added for generating and storing Consul tokens for EdgeX services to access Consul later on.

## Does this PR introduce a breaking change?


- [x ] No

## New Imports


- [x ] No

## Specific Instructions
Are there any specific instructions or things that should be known prior to reviewing?


## Other information
To test this locally, follows the steps below:
1. git clone this PR, build a local `dev` version of docker images via `make docker`.  
2. Then, in `edgex-compose`, under compose builder generate the compose file and update the docker-compose file to add `/tmp/edgex/secrets` on `consul` service as writable like:

```yaml
   consul:
      entrypoint: ["/edgex-init/consul_wait_install.sh"]
    .........
    volumes:
      - edgex-init:/edgex-init:ro,z
      # make secrets directories writable as Consul tokens are saved underneath this base directory
      - /tmp/edgex/secrets:/tmp/edgex/secrets:z
     .........
```
3. do a `docker-compose up`.   One should observe that consul tokens are generated underneath `/tmp/edgex/secrets` directory with their service-key as sub-folder something like this logging messages:
```
......
level=INFO ts=2021-03-30T16:13:01.589955906Z app=edgex-security-bootstrapper source=acltokens.go:166 msg="internal Consul ACL system is ready"
    2021-03-30T16:13:01.590Z [INFO]  agent: Synced check: check=AppService-rules-engine
level=INFO ts=2021-03-30T16:13:01.600544874Z app=edgex-security-bootstrapper source=acltokens.go:423 msg="successfully created a new registry token"
level=INFO ts=2021-03-30T16:13:01.600833693Z app=edgex-security-bootstrapper source=acltokens.go:316 msg="successfully created a new agent token"
    2021-03-30T16:13:01.603Z [INFO]  agent: Updated agent's ACL token: token=agent
level=INFO ts=2021-03-30T16:13:01.603826539Z app=edgex-security-bootstrapper source=acltokens.go:368 msg="agent token is set with type [agent]"
level=INFO ts=2021-03-30T16:13:01.603908166Z app=edgex-security-bootstrapper source=command.go:538 msg="successfully set up agent token into agent"
level=INFO ts=2021-03-30T16:13:01.604227728Z app=edgex-security-bootstrapper source=aclbootstrap.go:109 msg="bootstrap token is written to /tmp/edgex/secrets/consul-acl-token/bootstrap_token.json"
level=INFO ts=2021-03-30T16:13:01.604297113Z app=edgex-security-bootstrapper source=command.go:200 msg="SecretStoreClient created"
level=INFO ts=2021-03-30T16:13:01.604362357Z app=edgex-security-bootstrapper source=command.go:440 msg="successfully got unique role names, total size=5"
level=INFO ts=2021-03-30T16:13:01.625375988Z app=edgex-security-bootstrapper source=command.go:307 msg="successfully saved access token for service [edgex-core-command]"
level=INFO ts=2021-03-30T16:13:01.639097506Z app=edgex-security-bootstrapper source=command.go:307 msg="successfully saved access token for service [edgex-core-data]"
level=INFO ts=2021-03-30T16:13:01.65802044Z app=edgex-security-bootstrapper source=command.go:307 msg="successfully saved access token for service [edgex-core-metadata]"
level=INFO ts=2021-03-30T16:13:01.671018684Z app=edgex-security-bootstrapper source=command.go:307 msg="successfully saved access token for service [edgex-support-notifications]"
level=INFO ts=2021-03-30T16:13:01.681975965Z app=edgex-security-bootstrapper source=command.go:307 msg="successfully saved access token for service [edgex-support-scheduler]"
level=INFO ts=2021-03-30T16:13:01.682209615Z app=edgex-security-bootstrapper source=command.go:276 msg="successfully set up EdgeX access tokens"
level=INFO ts=2021-03-30T16:13:01.682468128Z app=edgex-security-bootstrapper source=command.go:177 msg="setupRegistryACL successfully done"

......

```
